src/Microsoft.DotNet.Watcher.Core/Internal/ProjectWatcher.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Watcher.Core.Internal
+{
+    public class ProjectWatcher : IDisposable
+    {
+        private readonly IProjectProvider _projectProvider;
+        private readonly IFileWatcher _fileWatcher;
+
+        private readonly string _rootProject;
+        private readonly bool _watchProjectJsonOnly;
+
+        private ISet<string> _watchedFiles;
+
+        public ProjectWatcher(
+            string projectToWatch,
+            bool watchProjectJsonOnly,
+            Func<IFileWatcher> fileWatcherFactory,
```

Comment:
It's a matter of who's disposing the object. If everything is encapsulated behind that factory, the `ProjectWatcher` has full control over the lifetime of the object. If I would pass in an IFileWatcher, this object couldn't know if it can safely dispose it